bloggies_backend/src/routes/stripe.ts

```diff
@@ -1,27 +1,140 @@
  * Returns session id, currently not checking if user logged in */
 stripeRouter.post("/create-checkout-session", async function (req: Request, res: Response, next: NextFunction) {
   try {
-    const {priceId}= req.body;
+    const { priceId } = req.body;
     const session = await Checkout.stripeCheckout(priceId);
-    res.status(201).send({sessionId: session.id})
+    res.status(201).send({ sessionId: session.id })
   } catch (err) {
     return next(err);
   }
 });
 
+
+/** POST create a customer for a user */
+stripeRouter.post("/create-customer", ensureLoggedIn, async function (req: Request, res: Response, next: NextFunction) {
+  const { user_id, email } = req.user;
+  const customer = await Checkout.stripeCreateCustomer(user_id, email);
+
+  await User.updateUser(user_id, { customer_id: customer.id });
+
+  return res.json({ customer });
+});
+
+/** POST create a subscription for a customer and save payment method  */
+stripeRouter.post("/create-subscription", ensureLoggedIn, async function (req: Request, res: Response, next: NextFunction) {
+  const { paymentMethodId, customerId } = req.body
+
+  try {
+    // save payment method info for a customer
+    await stripe.paymentMethods.attach(paymentMethodId, {
+      customer: customerId
+    });
+  } catch (err) {
+    const expressErr = new ExpressError(err, 402);
+    return next(expressErr);
+  }
+  // update the customer with the payment method
+  await stripe.customers.update(customerId, {
+    invoice_settings: {
+      default_payment_method: paymentMethodId
+    }
+  });
+
+  const subscription = await Checkout.stripeCreateSubscription(customerId);
+
+  await User.updateUser(req.user.user_id, { subscription_id: subscription.id });
+
+  return res.json({ subscription });
+});
+
 /** DELETE cancels user's Stripe subscription */
 stripeRouter.delete("/cancel-subscription", async function (req: Request, res: Response, next: NextFunction) {
   try {
-    const cancel = await Checkout.stripeSubscriptionCancel();
-    res.send(cancel);
-  } catch(err) {
+    const cancelledSubscription = await Checkout.stripeSubscriptionCancel(req.body.subscription_id);
+    res.send(cancelledSubscription);
+  } catch (err) {
     return next(err);
   }
+});
+
+/** POST update the customer with new payment method and assign it as the new default payment for subscription invoices */
+stripeRouter.post("/retry-invoice", async function (req: Request, res: Response, next: NextFunction) {
+  try {
+    const { customer_id, paymentMethodId, invoiceId } = req.body;
+
+    await stripe.paymentMethods.attach(paymentMethodId, {
+      customer: customer_id
+    });
+    // update the customer with the payment method
+    await stripe.customers.update(customer_id, {
+      invoice_settings: {
+        default_payment_method: paymentMethodId
+      }
+    });
+
+    const invoice = await stripe.invoices.retrieve(invoiceId, {
+      expand: ['payment_intent']
```

Comment:
are you checking the invoice here to see if it was reprocessed since you are adding a new payment method?

Reply:
Yes! This endpoint will be used when there is a subscription payment failure! you can find this exact same endpoint in the docs here: https://stripe.com/docs/billing/subscriptions/fixed-price#manage-subscription-payment-failure 

if I have understood the Stripe flow correctly, retrieving the invoice should be enough to re-attempt the payment charge with the new PaymentMethod!

Reply:
Here's more info about `invoices.retrieve` here https://stripe.com/docs/api/invoices/retrieve